Lib/test/test_threading.py

```diff
@@ -2103,6 +2103,26 @@ def test__all__(self):
         support.check__all__(self, threading, ('threading', '_thread'),
                              extra=extra, not_exported=not_exported)
 
+    def test_set_name(self):
+        try:
+            get_name = _thread._get_name
+            set_name = _thread.set_name
+        except AttributeError:
+            self.skipTest("need thread._get_name() and thread.set_name()")
+
+        work_name = None
+        def work():
+            nonlocal work_name
+            work_name = get_name()
+
+        # name not too long to fit into Linux 15 bytes limit
+        name = "CustomName"
+
+        thread = threading.Thread(target=work, name=name)
+        thread.start()
+        thread.join()
+        self.assertEqual(work_name, name)
+
```

Comment:
Could we exercise some non-ASCII edge cases?
```suggestion
        # name not too long to fit into Linux 15 bytes limit
        name = "Custom\N{SPOOL OF THREAD}"

        thread = threading.Thread(target=work, name=name)
        thread.start()
        thread.join()
        self.assertEqual(work_name, name)

        # name for which Linux truncation gives broken UTF-8
        name = "a" * 14 + "\N{SPOOL OF THREAD}" + "b" * 14

        thread = threading.Thread(target=work, name=name)
        thread.start()
        thread.join()
        self.assertTrue(work_name.startswith(name[:14]))

```